syntax = "proto3";

package testi;
// DRAW
message DrawPoint {
  double latitude = 1;
  double longitude = 2;
}

message PointArray{
  repeated DrawPoint points = 1;
}

message Fill{
  int32 color = 1;
}

message Stroke{
  int32 color = 1;
  int32 width = 2;
}

message Polygon{
  repeated PointArray pointarray = 1;
  int32 id = 2;
  Stroke stroke = 3;
  Fill fill = 4;
}

message Circle{
  DrawPoint center = 1;
  double radius = 2;
  int32 id = 3;
  Stroke stroke = 4;
  Fill fill = 5;
}
message Linestring {
  PointArray pointarray = 1;
  int32 id = 2;
  Stroke stroke = 3;
}

message DrawShape {
  int32 senderID = 1;
  repeated Linestring linestrings = 2;
  repeated Polygon polys = 3;
  repeated Circle circles = 4;
}

// DRAW END


//CHAT&& LOCATION
message Chatmessage {
  int32 senderID = 1;
  string msg = 2;
}

message Location {
  int32 senderID = 1;
  double latitude = 2;
  double longitude = 3;
  double accuracy = 4;
}
//CHAT&&LOCATION END

message FromServer {
  repeated Chatmessage chatmsg = 1;
  repeated Location locations = 2;
  repeated DrawShape shapes = 3;
  repeated NewGroup newgroups = 4;
  repeated EditGroup editgroups = 5;
  repeated UserMoved usermoved = 6;
  LoginAnswer loginanswer = 16;
  JoinRoomAnswer joinanswer = 17;
  string errmsg = 18;
}

message ToServer {
  Query query = 1;
  SendLoginInfo logininfo = 2;
  string chatmsg = 3;
  Location location = 4;
  DrawShape shape = 5;
  JoinRoom joinroom = 6;
  CreateGroup creategroup = 7;
  JoinGroup joingroup = 8;

  Fill fill = 11;
  Stroke stroke = 10;
  EditGroup editgroup = 9;
}

// GROUPS
message CreateGroup {
  string name = 1;
  uint32 color = 2;
}

message UserMoved {
  int32 userid = 1;
  int32 groupid = 2;
  string name = 3;
}

message JoinGroup {
  int32 id = 1;
}

message NewGroup {
  int32 id = 1;
  string name = 2;
  uint32 usercolor = 3;
}
message EditGroup {
  int32 id = 1;
  string name = 2;
  uint32 usercolor = 3;
}

// LOGIN&&JOIN

message LoginAnswer {
  string key = 1;
  string username = 2;
  bool success = 3;
  string errmsg = 4;
  string oldroom = 5;
}

message SendLoginInfo {
  string key = 1;
  string username = 2;
}

message JoinRoom {
  string roomname = 1;
  string password = 2;
  bool createroom = 3;
}

message JoinRoomAnswer {
  bool success = 1;
  bool wrongpassword = 2;
  bool roomexist = 3;
  string errmsg = 4;
  int32 id = 5;
}
// LOGIN&&JOIN END

message Query {
  uint64 cmd = 1;
  string args = 2;
}


